packages/tlschema/src/shapes/TLArrowShape.ts

```diff
@@ -123,16 +115,88 @@ export const arrowShapeMigrations = createShapePropsMigrationSequence({
 					delete end.isPrecise
 				}
 			},
-		},
+		}),
 
-		{
+		propsMigration({
 			id: arrowShapeVersions.AddLabelPosition,
 			up: (props) => {
 				props.labelPosition = 0.5
 			},
 			down: (props) => {
 				delete props.labelPosition
 			},
+		}),
+
+		{
+			id: arrowShapeVersions.ExtractBindings,
+			scope: 'store',
+			up: (oldStore) => {
+				type OldArrowTerminal =
+					| {
+							type: 'point'
+							x: number
+							y: number
+					  }
+					| {
+							type: 'binding'
+							boundShapeId: TLShapeId
+							normalizedAnchor: VecModel
+							isExact: boolean
+							isPrecise: boolean
+					  }
+					// new type:
+					| { type?: undefined; x: number; y: number }
+
+				type OldArrow = TLBaseShape<'arrow', { start: OldArrowTerminal; end: OldArrowTerminal }>
+
+				const arrows = Object.values(oldStore).filter(
+					(r: any): r is OldArrow => r.typeName === 'shape' && r.type === 'arrow'
+				)
+
+				for (const arrow of arrows) {
+					const { start, end } = arrow.props
+					if (start.type === 'binding') {
+						const id = createBindingId()
+						const binding: TLArrowBinding = {
+							typeName: 'binding',
+							id,
+							type: 'arrow',
+							fromId: arrow.id,
+							toId: start.boundShapeId,
+							meta: {},
+							props: {
+								terminal: 'start',
+								normalizedAnchor: start.normalizedAnchor,
+								isExact: start.isExact,
+								isPrecise: start.isPrecise,
+							},
+						}
+
+						oldStore[id] = binding
+						arrow.props.start = { x: 0, y: 0 }
+					}
```

Comment:
should this have an else clause that deletes the `type` property on the terminal?
```suggestion
					} else {
					  delete arrow.props.start.type
					}
```

same for the end binding below?